### PR TITLE
Visualizers that display basic aiida objects

### PR DESCRIPTION
**Pull request by yakutovicha**

### CODE REVIEW

File: aiidalab_widgets_base/aiida_visualizers.py [hunk 1]
```diff
@@ -0,0 +1,139 @@
+from __future__ import print_function
+import os
+
+import ipywidgets as ipw
+from aiida import load_dbenv, is_dbenv_loaded
+from aiida.backends import settings
+if not is_dbenv_loaded():
+    load_dbenv(profile=settings.AIIDADB_PROFILE)
```
Comment (ltalirz) [inline aiidalab_widgets_base/aiida_visualizers.py]: please remove

File: aiidalab_widgets_base/aiida_visualizers.py [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+
+class ParameterDataVisualizer(ipw.HTML):
+    """Basic visualizer class for ParameterData object"""
```
Comment (ltalirz) [inline aiidalab_widgets_base/aiida_visualizers.py]:
```suggestion
    """Visualizer class for ParameterData object"""
```

Comment (ltalirz) [reply aiidalab_widgets_base/aiida_visualizers.py]: similar for other docstrings

File: aiidalab_widgets_base/aiida_visualizers.py [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+            td { min-width: 300px; text-align: center; }
+        </style>
+        '''
+        pd.set_option('max_colwidth', 40)
+        df = pd.DataFrame([(key, value) for key, value
+                           in sorted(parameter.get_dict().items())
+                          ], columns=['Key', 'Value'])
+        self.value += '<center>' + df.to_html(classes='df', index=False) + '</center>'
+        if downloadable:
+            payload = base64.b64encode(df.to_csv().encode()).decode()
```
Comment (ltalirz) [inline aiidalab_widgets_base/aiida_visualizers.py]:
```suggestion
            payload = base64.b64encode(df.to_csv(index=False).encode()).decode()
```

File: aiidalab_widgets_base/aiida_visualizers.py [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+    def __init__(self, structure, downloadable=True, **kwargs):
+        import nglview
+        self._structure = structure
+        viewer = nglview.NGLWidget()
+        viewer.add_component(nglview.ASEStructure(self._structure.get_ase()))  # adds ball+stick
+        viewer.add_unitcell()
+        children = [viewer]
+        if downloadable:
+            self.file_format = ipw.Dropdown(
+                options=['xyz', 'cif', 'json'],
```
Comment (ltalirz) [inline aiidalab_widgets_base/aiida_visualizers.py]: what is this `json` format?
Is it some ASE-internal stuff?

File: aiida_datatypes.ipynb [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "FolderData = DataFactory('folder')\n",
+    "fd = FolderData()\n",
+    "with fd.folder.open(path.join('path','test1.xyz'), 'w') as fobj:\n",
```
Comment (ltalirz) [inline aiida_datatypes.ipynb]: perhaps rather use .txt extension since it's not actually an xyz.

File: aiidalab_widgets_base/display.py [hunk 6]
```diff
@@ -0,0 +1,22 @@
+from __future__ import print_function
+
+import importlib
+from IPython.display import display
+
+RECOGNIZED_AIIDA_DATA_VISUALIZERS = {
```
Comment (ltalirz) [inline aiidalab_widgets_base/display.py]: Perhaps `AIIDA_VISUALIZER_MAPPING`?

File: aiidalab_widgets_base/display.py [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+    'data.parameter.ParameterData.' : 'ParameterDataVisualizer',
+    'data.structure.StructureData.' : 'StructureDataVisualizer',
+    'data.cif.CifData.'             : 'StructureDataVisualizer',
+    'data.folder.FolderData.'       : 'FolderDataVisualizer',
+    'data.array.bands.BandsData.'   : 'BandsDataVisualizer',
+}
+
+def aiidalab_display(obj, downloadable=True):
+    """Function that is able to display properly differnt AiiDA data types"""
```
Comment (ltalirz) [inline aiidalab_widgets_base/display.py]:
```suggestion
    """Display AiiDA data types in Jupyter notebooks.
    
    :param downloadable: If True, add link/button to download content of displayed AiiDA object.
    
    Defers to IPython.display.display for any objects it does not recognize.
    """
```

File: aiidalab_widgets_base/display.py [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+
+RECOGNIZED_AIIDA_DATA_VISUALIZERS = {
+    'data.parameter.ParameterData.' : 'ParameterDataVisualizer',
+    'data.structure.StructureData.' : 'StructureDataVisualizer',
+    'data.cif.CifData.'             : 'StructureDataVisualizer',
+    'data.folder.FolderData.'       : 'FolderDataVisualizer',
+    'data.array.bands.BandsData.'   : 'BandsDataVisualizer',
+}
+
+def aiidalab_display(obj, downloadable=True):
```
Comment (ltalirz) [inline aiidalab_widgets_base/display.py]: kwargs?

File: aiidalab_widgets_base/display.py [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+    """Function that is able to display properly differnt AiiDA data types"""
+    from aiidalab_widgets_base import aiida_visualizers
+    if hasattr(obj, 'type') and obj.type in RECOGNIZED_AIIDA_DATA_VISUALIZERS:
```
Comment (ltalirz) [inline aiidalab_widgets_base/display.py]: The more pythonic way would be to just try:

```python
try:
    visualizer = getattr(aiida_visualizers, AIIDA_DATA_VISUALIZER_MAP[obj.type])
    display( visualizer(obj, downloadable=downloadable), **kwargs)
except KeyError:
    display(obj, **kwargs)
```

**Comment by ltalirz** [comment]
Anyhow, I would say: good job!
I think the overall implementation is quite lightweight and will certainly be useful.

Can you think a bit about how to reuse code between the "visualizers" and the "uploadwidgets"?

File: aiidalab_widgets_base/aiida_visualizers.py (hunk 1 above)
Comment (yakutovicha) [reply aiidalab_widgets_base/aiida_visualizers.py]: done

File: aiidalab_widgets_base/aiida_visualizers.py (hunk 2 above)
Comment (yakutovicha) [reply aiidalab_widgets_base/aiida_visualizers.py]: done

File: aiidalab_widgets_base/aiida_visualizers.py (hunk 3 above)
Comment (yakutovicha) [reply aiidalab_widgets_base/aiida_visualizers.py]: done

File: aiida_datatypes.ipynb (hunk 5 above)
Comment (yakutovicha) [reply aiida_datatypes.ipynb]: done

File: aiidalab_widgets_base/display.py (hunk 6 above)
Comment (yakutovicha) [reply aiidalab_widgets_base/display.py]: Sure. That is definitely a better name.

Thanks

File: aiidalab_widgets_base/display.py (hunk 7 above)
Comment (yakutovicha) [reply aiidalab_widgets_base/display.py]: Thanks, again.

File: aiidalab_widgets_base/display.py (hunk 8 above)
Comment (yakutovicha) [reply aiidalab_widgets_base/display.py]: right, fixed.

File: aiidalab_widgets_base/display.py (hunk 9 above)
Comment (yakutovicha) [reply aiidalab_widgets_base/display.py]: Really?

I am of course happy to change but are you sure `try-except` is any better in this case?

Comment (yakutovicha) [reply aiidalab_widgets_base/display.py]: Well, this is definitely shorter and more clear. I go for your advice.

Comment (ltalirz) [reply aiidalab_widgets_base/display.py]: > shorter and more clear

That was also my thinking, following the "EAFP" motto.

Even if there might be a slight penalty (not sure there is), you are just trying once per visualizer so there is no point in worrying about it.

File: aiidalab_widgets_base/aiida_visualizers.py (hunk 4 above)
Comment (yakutovicha) [reply aiidalab_widgets_base/aiida_visualizers.py]: Yes, contains all the information that is present in ASE `Atoms` object. [Here](https://wiki.fysik.dtu.dk/ase/_modules/ase/io/formats.html) they just say "ASE JSON database file".

Might be useful if the structure is going to be converted into ASE again.

**Comment by yakutovicha** [comment]
> I find the table styling not particularly nice. pandas provides much nicer formatting out of the box.

I have a mid-way solution here. I still prefer that widget inherits from ipw.HTML (to keep traitlets). However, this kills all the formatting that pandas does. To make everyone happy I adapted my CSS styling to what pandas does.

Please have a look.

**Comment by yakutovicha** [comment]
> Can you think a bit about how to reuse code between the "visualizers" and the "uploadwidgets"?

Sure, I will make a new issue for that

File: aiidalab_widgets_base/aiida_visualizers.py (hunk 4 above)
Comment (ltalirz) [reply aiidalab_widgets_base/aiida_visualizers.py]: Ok. Have you ever used it?
I believe most people won't know what "json" stands for here.
To keep things simple, perhaps better remove it.

File: aiidalab_widgets_base/aiida_visualizers.py [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+import os
+
+import ipywidgets as ipw
+
+class ParameterDataVisualizer(ipw.HTML):
+    """Visualizer class for ParameterData object"""
+    def __init__(self, parameter, downloadable=True, **kwargs):
+        super(ParameterDataVisualizer, self).__init__(**kwargs)
+        import pandas as pd
+        self.value = '''
```
Comment (ltalirz) [inline aiidalab_widgets_base/aiida_visualizers.py]: Could you add a comment perhaps with a link here, explaining where the CSS is coming from?

File: aiidalab_widgets_base/aiida_visualizers.py (hunk 4 above)
Comment (yakutovicha) [reply aiidalab_widgets_base/aiida_visualizers.py]: ok

File: aiidalab_widgets_base/aiida_visualizers.py (hunk 10 above)
Comment (yakutovicha) [reply aiidalab_widgets_base/aiida_visualizers.py]: Sure.

Of course "the" link does not exist. I was basically acting on the standard html table settings.

I added a couple of lines describing what pandas exported object is, and how to control its appearance through css styling.

**Pull request merged**